lib/framedivider.rb

```diff
@@ -0,0 +1,305 @@
+#
+# @param [Array] pts Point3D array (3x)
+#
+# @return [Float] triangle area (m2)
+# @return [Array] subsurface rough opening sorted Topolys 3D points
+def areaHeron(pts)
+  raise "Invalid pts (areaHeron)" unless pts
+  raise "#{pts.class}? expected Array (areaHeron)" unless pts.is_a?(Array)
+  raise "3x points please - got #{pts.size} (areaHeron)" unless pts.size == 3
+
+  cl = Topolys::Point3D
+  pts.each do |pt|
+    raise "Topolys 3D points = got #{pt.class}" unless pt.class == cl
+  end
+
+  e = []
+  e << { v0: pts[0], v1: pts[1], mag: (pts[1] - pts[0]).magnitude }
+  e << { v0: pts[1], v1: pts[2], mag: (pts[2] - pts[1]).magnitude }
+  e << { v0: pts[2], v1: pts[0], mag: (pts[0] - pts[2]).magnitude }
+
+  raise "Unique triangle edges (areaHeron)" if matches?(e[0], e[1])
+  raise "Unique triangle edges (areaHeron)" if matches?(e[1], e[2])
+  raise "Unique triangle edges (areaHeron)" if matches?(e[2], e[0])
+
+  e = e.sort_by{ |p| p[:mag] }
+
+  area = 0
+  # Kahan's stable implementation of Heron's formula for triangle area.
+  # area = 1/4 sqrt( (a+(b+c)) (c-(a-b)) (c+(a-b)) (a+(b-c)) ) ... a > b > c
+  a = e[0][:mag]
+  b = e[1][:mag]
+  c = e[2][:mag]
+  return 0.25 * Math.sqrt( (a+(b+c)) * (c-(a-b)) * (c+(a-b)) * (a+(b-c)) )
+end
+
+##
+# Calculate subsurface rough opening.
+#
+# @param [OpenStudio::Model::Model] model An OS model
+# @param [String] id SubSurface identifier
+# @param [] id SubSurface identifier
+#
+# @return [Float] subsurface rough opening area (m2)
+# @return [Array] subsurface rough opening sorted Topolys 3D points
+def opening(model, id, t)
+  raise "Invalid model (roughOpening)" unless model
+  raise "Invalid ID (roughOpening)" unless id
+
+  cl = OpenStudio::Model::Model
+  raise "#{model.class}? expected #{cl} (roughOpening)" unless model.is_a?(cl)
+
+  s = model.getSubSurfaceByName(id)
+  raise "#{id} SubSurface missmatch!" if s.empty?
+  s = s.get
+
+  raise "Invalid parent surface (roughOpening)" if s.surface.empty?
+  parent = model.getSurfaceByName(s.surface.get.nameString)
+  raise "#{id} missing parent surface (roughOpening)!" if parent.empty?
+  parent = parent.get
+
+  area = s.grossArea
+  raise "#{id} gross area < 0 m2 (roughOpening)?" unless area > TOL
+
+  n = s.vertices.size
+  raise "#{id} vertex count, 3 or 4 (roughOpening)!" unless n == 3 || n == 4
+
+  dad = (t * parent.vertices).map{ |v| Topolys::Point3D.new(v.x, v.y, v.z) }
+  points = (t * s.vertices).map{ |v| Topolys::Point3D.new(v.x, v.y, v.z) }
+
+  # Should verify convexity of vertex wire/face ...
+  #
+  #       A
+  #      / \
+  #     /   \
+  #    /     \
+  #   / C --- D    <<< allowed as OpenStudio/E+ subsurface?
+  #  / /
+  #  B
+  #
+  # Should convert (annoying) 4-point subsurface into triangle ...
+  #        A
+  #       / \
+  #      /   \
+  #     /     \
+  #    B - C - D   <<< allowed as OpenStudio/E+ subsurface?
+  #
+  width = 0
+  if s.windowPropertyFrameAndDivider.empty?
+    return area, points
```

Comment:
No Frame & Divider object? Return original subsurface gross area (or net, same thing) and _transformed_ (otherwise unaltered) points.